packages/nimble-components/src/table/specs/table-column-specs/table-column-mapping.md

```diff
@@ -0,0 +1,306 @@
+
+-   `nimble-mapping-string` (maps string values)
+-   `nimble-mapping-number` (maps number values)
+-   `nimble-mapping-boolean` (maps boolean values)
+
+_Props/Attrs_
+
+-   `key`: string (property type matches component type, e.g. `number` for number mapping)
+-   `description`: string
+
+_Content_
+
+-   exactly one `template` element, containg some combination of text, Nimble icons, or `nimble-spinner`
+
+### Anatomy
+
+#### `nimble-table-column-mapping`
+
+```HTML
+<template slot="${x => x.columnInternals.uniqueId}">
+    <slot
+        ${slotted('mappings')}
+        name="mapping">
+    </slot>
+    <span class="header-content">
+        <slot></slot>
+    </span>
+</template>
+```
+
+Cell view:
+
+```HTML
+${repeat(x => (x.column as TableColumnMapping).mappings,
+    html<TableColumnMapping, TableColumnMappingCellView>`
+        ${when((x, c) => x.key === (c.parent as TableColumnMappingCellView).cellRecord.value,
+            html<TableColumnMapping>`
+                ${repeat(x => x.mappedTemplate,
+                    html<HTMLTemplateElement>`${x => new ViewTemplate(x.innerHTML, [])}`
+                )}
+            `
+        )}
+    `
+)}
+```
+
+Header view:
+
+Note the following requires that `TableColumnMappingHeaderView` has a reference to the column with which it is associated. This is needed to enumerate the column's mapping elements.
+
+```HTML
+${repeat(x => (x.column as TableColumnMapping).mappings,
+    html<TableColumnMapping, TableColumnMappingHeaderView>`
+        ${when((x, c) => x.key === (c.parent as TableColumnMappingHeaderView).groupHeaderValue,
+            html<TableColumnMapping>`
+                ${repeat(x => x.mappedTemplate,
+                    html<HTMLTemplateElement>`${x => new ViewTemplate(x.innerHTML, [])}`
+                )}
+            `
+        )}
+    `
+)}
+```
+
+For each mapping, if the key matches the value for this row, get the mapping's contained `template` and render its `innerHTML`.
+
+#### `nimble-mapping-string/number/boolean`
+
+```HTML
+<template slot="mapping">
+    <slot
+        ${slotted({ property: 'mappedTemplate', filter: x => x instanceof HTMLTemplateElement })}
+    >
+    </slot>
+</template>
+```
+
+### Grouping
+
+Will support grouping by the record value. The grouping header will display the rendered html mapped to that value. If multiple different values map to the same rendered html (e.g. icon), there may be separate groups that visually look like they should be combined into one group. We could instead render the record value into the header, but that value is not otherwise exposed to the user. It may not even have much meaning (e.g. some numeric code). I think it is more reasonable to display the rendered html.
+
+### Sorting
+
+Sorting has the same issue as grouping, where visually identical cells could end up not sorted contiguously. Again, I believe this is reasonable.
+
+### Sizing
+
+We will support fixed or fractional widths for this column type.
```

Comment:
To better support mapping columns with only icon mappings, should we export a constant (or a token?) that makes it easy for a client to create an icon column with the recommended fixed size?

Reply:
I'm just now looking at how we support fixed-width columns. Do I need to define a `pixel-width` attribute on the column type and sync that with `columnInternals.currentPixelWidth`? The example in the spec shows setting `columnInternals.currentPixelWidth` in the component's constructor, but that doesn't seem applicable to this situation. But to answer your question, I suppose it would be helpful to export a constant for a single-icon column width.

Reply:
So it sounds like we need to define a new mixin to expose the public API for setting fixed widths on columns. It would do basically what I suggested in my last comment. Also, for icon-only columns, it sounds like the default minimum column width is significantly more than the width of an icon. So maybe we don't need to have a token for the column width. We can just document to set the width to 1 and let the minimum govern the width.

Reply:
I'm not entirely sure what you mean by "set the width to 1 and let the minimum govern the width". Just so we're on the same page about things, the default behavior for _any_ column, regardless of the mixins they have applied, will be to behave as a fractional width column of `1fr`. If we want an icon column to default to a pixel width behavior, then it is _required_ to set the `currentPixelWidth` to a non-undefined value.

This is arguably a problem with how the current API is designed. That is, if we had something like a `FixedWidthColumn` mixin that itself did _not_ extend the `FractionalWidthColumn`, then unless a non-null default was provided for the `pixelWidth` (which perhaps it should be given the same default as `minPixelWidth`), then it would still behave like a fractional width column by default.

Reply:
I had not assumed we should make this column type use a fixed width by default, but I can see the argument for doing so. It could result in truncated text (by default) for columns that map values to text, but it's arguably better to expect clients to override the default in those cases than have to set a fixed width in all the other cases (i.e. with just icons/spinners).

Reply:
Now we're going to have a separate column type for icons which only supports fixed width. The other, general mapping type will support both, but default to fractional.